lib/msal-node/src/client/ManagedIdentitySources/BaseManagedIdentitySource.ts

```diff
@@ -57,8 +60,27 @@ export abstract class BaseManagedIdentitySource {
     ) {
         this.logger = logger;
         this.nodeStorage = nodeStorage;
-        this.networkClient = networkClient;
         this.cryptoProvider = cryptoProvider;
+
+        const maxRetries = 1;
+        const retryDelay = 1000;
+        const httpStatusCodesToRetryOn = [
+            HttpStatus.NOT_FOUND,
+            HttpStatus.REQUEST_TIMEOUT,
+            HttpStatus.TOO_MANY_REQUESTS,
+            HttpStatus.INTERNAL_SERVER_ERROR,
+            HttpStatus.SERVICE_UNAVAILABLE,
+            HttpStatus.SERVER_ERROR_RANGE_END,
+        ];
+        const linearRetryPolicy: LinearRetryPolicy = new LinearRetryPolicy(
+            maxRetries,
+            retryDelay,
+            httpStatusCodesToRetryOn
+        );
+        this.networkClientWithRetry = new HttpClientWithRetries(
```

Comment:
What is the lifetime of BaseManagedIdentitySource object in relation to token requests? 
Will each token request have create a `BaseManagedIdentitySource` or is this object reused by all requests? If it's reused, then the code is wrong, because the HttpClient retains the "current retry" as state.

Irrespective of the answer, there needs to be a test for this: 

- create a ManagedIdentityApp 
- configure the HTTP stack so that it always returns 500
- request a token for resource R1. Assert that 2 HTTP calls are made.
- request a token for resource R2. Assert that 2 HTTP calls are made.

Reply:
The lifetime of BaseManagedIdentitySource object is once per initial request (used by all retries). It's not used across all requests.

I will create the tests you suggested to confirm this.

Reply:
I've been looking into this.

The `BaseManagedIdentitySource` is the static `identitySource` in `ManagedIdentityClient`, with the `ManagedIdentityClient` created per `ManagedIdentityApplication`. So `BaseManagedIdentitySource` (`identitySource`) is defined upon the first network request of each application, and is used across applications (This is desired behavior per our previous discussions and current implementation).

So according to your comments, this makes the code in this PR wrong. What do you mean by that, can you elaborate? Should the `networkClientWithRetry` be defined in `ManagedIdentityApplication` before `acquireToken` is called?

Reply:
I think you just need to write a unit test and you'll figure out why it's wrong. The unit test should use only public API and HTTP injection.


1. create a ManagedIdentity setup (public API only)
2. configure the HTTP stack so that it always returns 500
3. request a token for resource R1. 

**Expected**: Assert that 2 HTTP calls are made.
4. request a token for resource R2. Assert that 2 HTTP calls are made.
**Expected**: Assert that 2 HTTP calls are made.

**Actual**: I believe in your case, only 1 HTTP call will be made here.

Reply:
I wrote a unit test for this scenario:

```
test("creates two different managed identity applications (that use the same HttpClient) which both return 500 error responses from their network requests permanently, to ensure that retry policy lifetime is per request", async () => {
                expect(ManagedIdentityTestUtils.isIMDS()).toBe(true);

                const networkClient2: ManagedIdentityNetworkClient =
                    new ManagedIdentityNetworkClient(
                        MANAGED_IDENTITY_RESOURCE_ID_2 // client id
                    );
                const managedIdentityApplication2: ManagedIdentityApplication =
                    new ManagedIdentityApplication({
                        system: {
                            networkClient: networkClient2,
                            // managedIdentityIdParams will be omitted for system assigned
                        },
                    });

                const sendGetRequestAsyncSpyApp1: jest.SpyInstance = jest
                    .spyOn(networkClient, <any>"sendGetRequestAsync")
                    // permanently override the networkClient's sendGetRequestAsync method to return a 500
                    .mockReturnValue(
                        managedIdentityNetworkErrorClient.sendGetRequestForRetryAsync()
                    );

                const sendGetRequestAsyncSpyApp2: jest.SpyInstance = jest
                    .spyOn(networkClient2, <any>"sendGetRequestAsync")
                    // permanently override the networkClient's sendGetRequestAsync method to return a 500
                    .mockReturnValue(
                        managedIdentityNetworkErrorClient.sendGetRequestForRetryAsync()
                    );

                try {
                    await managedIdentityApplication.acquireToken(
                        managedIdentityRequestParams
                    );
                } catch (e) {
                    expect(sendGetRequestAsyncSpyApp1).toHaveBeenCalledTimes(2);
                    expect(sendGetRequestAsyncSpyApp2).toHaveBeenCalledTimes(0);
                }

                try {
                    await managedIdentityApplication2.acquireToken(
                        managedIdentityRequestParams
                    );
                } catch (e) {
                    expect(sendGetRequestAsyncSpyApp1).toHaveBeenCalledTimes(2);
                    expect(sendGetRequestAsyncSpyApp2).toHaveBeenCalledTimes(2);
                }

                jest.restoreAllMocks();
            });
```

It fails in the 2nd catch, on the line ```expect(sendGetRequestAsyncSpyApp1).toHaveBeenCalledTimes(2);```. The spy is being called 4 times, instead of 2.

Since ```identitySource``` being static is desired behavior, then what is your proposed solution here? 

Should the `networkClientWithRetry` be defined in `ManagedIdentityApplication` inside of each `acquireToken` call? Add the retryPolicy to the `ManagedIdentityRequest` inside of `acquireToken`?

Reply:
It feels that you are unsure of the object lifetime for all these objects that you've created. And I think the unit tests don't help in understanding, because they are quite complex. 

Here's a suggestion to help with understanding - create a new project and add the MSI classes to it, but with very simplified logic. No async, no helpers etc. Just constructor and public methods and return dumb values. Something like this: https://gist.github.com/bgavrilMS/fd70a44bca9ad194d2d453c4e7b3669d

And then run everything with a debugger and understand the lifetime of each object, how the interact with each other etc.

Reply:
Also, this is not the test I asked to write (but it's a good test as well). 

I asked for a test where you use a single `ManagedIdentityNetworkClient` object


```ts
 try {
                    await managedIdentityApplication.acquireToken(
                        managedIdentityRequestParams1 // resource 1
                    );
                } catch (e) {
                    expect(sendGetRequestAsyncSpyApp).toHaveBeenCalledTimes(2);                 
                }

                try {
                    await managedIdentityApplication.acquireToken(
                        managedIdentityRequestParams2 // resource 2
                    );
                } catch (e) {
                    expect(sendGetRequestAsyncSpyApp).toHaveBeenCalledTimes(2);                 
                }
```

Reply:
Spoke with Bogdan offline about this and cleared up confusion on both sides. Current implementation works based on new unit test. No change is needed.